File: Leetcode_Java/FindPeakElement.java
public class FindPeakElement {
    public int findPeakElement(int[] nums) {
        if (nums.length == 1) {
            return 0;
        }
        int start = 0;
        int end = nums.length - 1;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] > nums[mid - 1] && nums[mid] > nums[mid + 1]) {
                return mid;
            } else if (nums[mid] > nums[mid - 1]) {
                start = mid;
            } else {
                end = mid;
            }
        }
        return nums[start] > nums[end] ? start : end;
    }

    public static void main(String[] args) {
        FindPeakElement test = new FindPeakElement();
        int[] nums = {1, 2, 0};
        System.out.println(test.findPeakElement(nums));
    }
}

File: Leetcode_Python/find_peak_element.py
'''Version 1: Binary Search O(logN)'''
class Solution:
    # @param num, a list of integer
    # @return an integer
    def findPeakElement(self, num):
        left = 0; right = len(num) - 1
        while left <= right:
            if left == right:
                return left
            mid = (left + right) / 2
            if num[mid] > num[mid + 1]:
                right = mid
            else:
                left = mid + 1

'''Version 2: Brute-Force O(n)'''
class Solution:
    def findPeakElement(self, num):
        if len(num) == 1:   return 0
        if num[-1] > num[-2]:   return len(num) - 1
        i = 0
        while i <= len(num) - 2:
            if num[i] < num[i + 1]:
                i += 1
            else:
                return i

